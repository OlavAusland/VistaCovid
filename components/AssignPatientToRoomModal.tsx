import { Text, View, TextInput, Pressable, Modal, TouchableOpacity } from 'react-native';
import React, { useEffect, useState } from 'react';
import { getPatient } from '../api/folkeregisterModelAPI';
import { FolkeregisterPerson } from '../domain/PatientType';
import { assignPatientStyle } from '../styles/AssignPatientStyle';
import { dropdownStyles } from '../styles/dropdownStyle';
import { getAvailableRooms } from '../api/firebaseAPI';
import { Room } from '../domain/RoomType';
import { Dropdown } from 'react-native-element-dropdown';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import {addPatientToRoom} from '../api/firebaseAPI';



type AssignPatientModalProps = {
    modalVisible: boolean;
    handleRequestClose: Function;
}

type DropDownType = {
    open: boolean,
    value: string,
    items: Array<ItemType>
    room: string
}

type ItemType = {
    label: string,
    value: string
}

export const AssignPatientModal = (props: AssignPatientModalProps) => {

    const [patient, setPatient] = useState<FolkeregisterPerson>();
    const [search, setSearch] = useState<string>("");
    const [dropdown, setDropdown] = useState<DropDownType>({ open: false, value: "0", items: [], room: "" });

    const handleSearch = () => {
        if (search.length > 0) {
            const fnrRegex = new RegExp(/^(0[1-9]|[1-2][0-9]|31(?!(?:0[2469]|11))|30(?!02))(0[1-9]|1[0-2])\d{7}$/g);
            const isFnr = fnrRegex.test(search);

            if (isFnr) {
                getPatient(search).then(result => {
                    setPatient(result)
                }).catch(err => { console.log('here',err.message) });
            }
        }
    }

    const handleAddPatient = () => {
        if (patient && dropdown.room) {
            addPatientToRoom(dropdown.room, patient.ssn);
            props.handleRequestClose();
            setPatient(undefined);
            setSearch("");
            setDropdown({ open: false, value: "0", items: [], room: "" });
        }
    }


    useEffect(() => {
        console.log('here')
        getAvailableRooms().then((room: Room[]) => {
            setDropdown(prev => ({ ...prev, items: [] }));
            room.forEach((room: Room) => {
                const item: ItemType = { label: room.roomNumber, value: room.id, };
                setDropdown(prev => ({ ...prev, items: [...prev.items, item] }));
            })
        })
    }, [props.modalVisible]);

   

    return (

        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            visible={props.modalVisible}
            onRequestClose={() => props.handleRequestClose()}
        >
            <View style={{ top: '5%', }}>
                <View style={assignPatientStyle.container}>
                    <Text style={{ fontSize: 40, marginBottom: 90 }}>Admit patient</Text>
                    <Text style={{ fontSize: 20, paddingBottom: 5, marginTop: 40 }}>Patient:</Text>
                    <View style={{ width: '100%', borderRadius: 5, flexDirection: 'row', marginBottom: 10 }}>
                        <TextInput onChangeText={text => { setSearch(text) }} placeholder="SSN" style={{ flex: 3, height: 40, backgroundColor: "white", paddingLeft: 10, width: '100%' }} />
                        <TouchableOpacity style={{ flex: 1 }} onPress={() => handleSearch()}>
                            <View style={{ width: 70, backgroundColor: '#0274a1', height: 40, borderRadius: 10, marginLeft: 10 }}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 15, marginTop: 10 }}>Search</Text>
                            </View>
                        </TouchableOpacity>
                    </View>
                    {patient &&
                        <View>
                            <Text style={{ fontSize: 15 }}>{patient.lastname}, {patient.firstname} {patient.midlename}</Text>
                        </View>}
                    <Text style={{ fontSize: 20, paddingBottom: 5, marginTop: 25 }}>Room:</Text>
                    <View style={{ backgroundColor: 'white' }}>
                        <Dropdown
                            style={dropdownStyles.dropdown}
                            placeholderStyle={dropdownStyles.placeholderStyle}
                            selectedTextStyle={dropdownStyles.selectedTextStyle}
                            inputSearchStyle={dropdownStyles.inputSearchStyle}
                            iconStyle={dropdownStyles.iconStyle}
                            data={dropdown.items}
                            search
                            maxHeight={200}
                            labelField="label"
                            valueField="value"
                            placeholder="Select room"
                            searchPlaceholder="Search..."
                            value={dropdown.room}
                            onChange={item => {
                                setDropdown(prev => ({ ...prev, room: item.value }));
                            }}
                            renderLeftIcon={() => (
                                <FontAwesome5 style={dropdownStyles.icon} color="black" name="bed" size={20} />
                            )}
                        />
                    </View>
                    <View style={{ flexDirection: 'row', alignContent: 'space-between', marginTop: 200 }}>
                        <Pressable onPress={() => handleAddPatient()} >
                            <View style={assignPatientStyle.button}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Add</Text>
                            </View>
                        </Pressable>
                        <TouchableOpacity
                            onPress={() => props.handleRequestClose()} >
                            <View style={assignPatientStyle.button}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Cancel</Text>
                            </View>
                        </TouchableOpacity>
                    </View>
                </View>
            </View>
        </Modal>
    );
}

