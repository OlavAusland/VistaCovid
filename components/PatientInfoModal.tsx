import { Text, View, TextInput, Pressable, Modal, ScrollView } from 'react-native';
import React, { useEffect, useState } from 'react';
import { getPatient } from '../api/folkeregisterModelAPI';
import { FolkeregisterPerson } from '../domain/PatientType';
import { patientInfoStyle } from '../styles/PatientInfoStyle';


type PatientInfoModalProps = {
    modalVisible: boolean;
    handleRequestClose: Function;
    fnr: string;
}


export const PatientInfoModal = (props: PatientInfoModalProps) => {
    const [patient, setPatient] = useState<FolkeregisterPerson>();

    useEffect(() => {
        getPatient(props.fnr).then(patient => {
            setPatient(patient);
        }).catch(err => { console.log(err) });
    }, []);

    return (

        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            visible={props.modalVisible}
            onRequestClose={() => props.handleRequestClose()}
        >
            <View style={{ top: '20%', bottom: '15%' }}>
                <View style={patientInfoStyle.container}>
                    <Text style={{ fontSize: 30, fontWeight: '200', marginTop: 10, marginBottom: 10 }}>Patient</Text>
                    <View style={patientInfoStyle.text}>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Lastname: </Text>
                            <Text>{patient?.lastname}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Firstname: </Text>
                            <Text>{patient?.firstname} {patient?.midlename}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>SSN: </Text>
                            <Text>{patient?.ssn}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Gender: </Text>
                            <Text>{patient?.gender}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Address: </Text>
                            <Text>{patient?.address} {patient?.housenumber}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>City: </Text>
                            <Text>{patient?.city}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Co-Address: </Text>
                            <Text>{patient?.coAddress}</Text>
                        </Text>
                    </View>
                    <Pressable
                        onPress={() => props.handleRequestClose()} >
                        <View style={patientInfoStyle.button}>
                            <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Close</Text>
                        </View>
                    </Pressable>
                </View>
            </View>
        </Modal>
    );
}

