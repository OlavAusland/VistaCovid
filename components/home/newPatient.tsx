import { Text, View, TextInput, Pressable, Modal, TouchableOpacity } from 'react-native';
import { FolkeregisterPerson } from '../../domain/PatientType';
import { assignPatientStyle } from '../../styles/AssignPatientStyle';
import { dropdownStyles } from '../../styles/dropdownStyle';
import { Dropdown } from 'react-native-element-dropdown';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import { DropDownType } from '../../domain/DropDownType';
import { newPatientProps } from '../../domain/AssignPatietTypes';
import { auth } from '../../firebase-config';



export const NewPatient = (props: newPatientProps) => {
    return (
        <View>
            <View>
                <Text style={{ fontSize: 20, paddingBottom: 5 }}>Patient:</Text>
                    <TextInput onChangeText={text => props.setPatient((prev:FolkeregisterPerson) => ({...prev, ssn:text}))} placeholder="SSN" style={assignPatientStyle.newpatientinput}/>
                    <TextInput onChangeText={text => props.setPatient((prev:FolkeregisterPerson) => ({...prev, firstname:text}))} placeholder="Firstname" style={assignPatientStyle.newpatientinput}/>
                    <TextInput onChangeText={text => props.setPatient((prev:FolkeregisterPerson) => ({...prev, lastname:text}))} placeholder="Lastname" style={assignPatientStyle.newpatientinput}/>
                    <TextInput onChangeText={text => props.setPatient((prev:FolkeregisterPerson) => ({...prev, gender:text}))} placeholder="sex" style={assignPatientStyle.newpatientinput}/>
                    
                <Text style={{ fontSize: 20, paddingBottom: 5, marginTop: 25 }}>Room:</Text>
                <View style={{ backgroundColor: 'white' }}>
                    <Dropdown
                        style={dropdownStyles.dropdown}
                        placeholderStyle={dropdownStyles.placeholderStyle}
                        selectedTextStyle={dropdownStyles.selectedTextStyle}
                        inputSearchStyle={dropdownStyles.inputSearchStyle}
                        iconStyle={dropdownStyles.iconStyle}
                        data={props.dropdown.items}
                        search
                        maxHeight={200}
                        labelField="label"
                        valueField="value"
                        placeholder="Select room"
                        searchPlaceholder="Search..."
                        value={props.dropdown.label}
                        onChange={item => {
                            props.setDropdown((prev: DropDownType)=> ({ ...prev, label: item.value }));
                        }}
                        renderLeftIcon={() => (
                            <FontAwesome5 style={dropdownStyles.icon} color="black" name="bed" size={20} />
                        )}
                    />
                </View>
            </View>
            <View style={{ marginTop: 10 }}>
                        <Text>Responsible: {auth.currentUser?.displayName}</Text>
                    </View>
                    <View style={{ flexDirection: 'row', alignContent: 'space-between', marginTop: 140 }}>
                        <Pressable onPress={() => props.handleNewPatient()} >
                            <View style={assignPatientStyle.button}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Add</Text>
                            </View>
                        </Pressable>
                        <TouchableOpacity
                            onPress={() => {props.setNew(false); props.setError(''); props.setPatient(undefined) }} >
                            <View style={assignPatientStyle.button}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Cancel</Text>
                            </View>
                        </TouchableOpacity>
                    </View>
        </View>
    )


}