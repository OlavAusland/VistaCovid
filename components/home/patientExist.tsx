import { Pressable, Text, TextInput, TouchableOpacity, View } from "react-native";
import { Dropdown } from "react-native-element-dropdown";
import FontAwesome5 from "react-native-vector-icons/FontAwesome5";
import { existingPatientProps } from "../../domain/AssignPatietTypes";
import { DropDownType } from "../../domain/DropDownType";
import { auth } from "../../firebase-config";
import { assignPatientStyle } from "../../styles/AssignPatientStyle";
import { dropdownStyles } from "../../styles/dropdownStyle";


export const ExistingPatient = (props: existingPatientProps) => {
    return (
        <View>
            <View>
                <Text style={assignPatientStyle.existingPatient}> Patient: </Text>
                <View style={assignPatientStyle.existinginputcontainer}>
                    <TextInput
                        onChangeText={(text) => {props.setSearch(text); }}
                        placeholder="SSN"
                        style={assignPatientStyle.existingPatientInput} />
                    <TouchableOpacity
                        style={{ flex: 1 }}
                        onPress={() => props.handleSearch()}>
                        <View style={assignPatientStyle.existingsearch} >
                            <Text style={assignPatientStyle.existingSearchText}>Search </Text>
                        </View>
                    </TouchableOpacity>
                </View>
                {props.patient?.firstname && (
                    <View>
                        <Text style={assignPatientStyle.existngDisplay}>
                            {props.patient.lastname}, {props.patient.firstname}{" "}
                            {props.patient.midlename}
                        </Text>
                    </View>
                )}
                {props.error?.length > 0 && (
                    <View>
                        <Text style={assignPatientStyle.existngDisplay}>No patient Found</Text>
                        <Pressable onPress={() => props.handleNew()}>
                            <View style={assignPatientStyle.newPatientButton}>
                                <FontAwesome5 style={assignPatientStyle.plussIcon} color="white" name="plus" size={20}/>
                                <Text style={assignPatientStyle.addNewPatientButtonText}> New Patient </Text>
                            </View>
                        </Pressable>
                    </View>
                )}
                <Text style={assignPatientStyle.roomtext}> Room:</Text>
                <View style={assignPatientStyle.dropdownContainer}>
                    <Dropdown
                        style={dropdownStyles.dropdown}
                        placeholderStyle={dropdownStyles.placeholderStyle}
                        selectedTextStyle={dropdownStyles.selectedTextStyle}
                        inputSearchStyle={dropdownStyles.inputSearchStyle}
                        iconStyle={dropdownStyles.iconStyle}
                        data={props.dropdown.items}
                        search
                        maxHeight={200}
                        labelField="label"
                        valueField="value"
                        placeholder="Select room"
                        searchPlaceholder="Search..."
                        value={props.dropdown.label}
                        onChange={(item) => {
                            props.setDropdown((prev: DropDownType) => ({
                                ...prev,
                                label: item.value,
                            }));
                        }}
                        renderLeftIcon={() => (
                            <FontAwesome5 style={dropdownStyles.icon} color="black" name="bed" size={20}/>
                        )}
                    />
                </View>
            </View>
            <View style={{ marginTop: 10 }}>
                <Text> Responsible: {auth.currentUser ? auth.currentUser.displayName : 'User not found'}</Text>
            </View>
            <View
                style={assignPatientStyle.existingFooter}>
                <Pressable onPress={() => props.handleAddPatient()}>
                    <View style={assignPatientStyle.button}>
                        <Text style={assignPatientStyle.buttontext}>Add</Text>
                    </View>
                </Pressable>
                <TouchableOpacity onPress={() => props.setEmpty()}>
                    <View style={assignPatientStyle.button}>
                        <Text style={assignPatientStyle.buttontext}>Cancel</Text>
                    </View>
                </TouchableOpacity>
            </View>
        </View>
    );
};
