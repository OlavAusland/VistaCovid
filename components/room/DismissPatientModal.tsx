import { Modal, Pressable, Text, View } from "react-native";
import { adminStyle } from "../../styles/AdminStyles";

type DismissPatientModal={
    modalVisible: boolean;
    handleRequestClose: Function;
    dismissAndClose: Function;
}

export const DismissPatientModal = (props: DismissPatientModal) => {
    
    return(
        <Modal
        animationType="slide"
        statusBarTranslucent={true}
        transparent={true}
        visible={props.modalVisible}
        onRequestClose={()=>props.handleRequestClose()}>

                <View style={adminStyle.centeredView}>
                    <View style={adminStyle.modalView}>
                        <Text style={adminStyle.modalText}>Are you sure you want to dismiss patient?</Text>
                        
                        <Pressable
                        style={[adminStyle.button, adminStyle.buttonOpen]}
                        onPress={()=> props.dismissAndClose()}
                        >
                            <Text style={adminStyle.textStyle}>Yes, dismiss.</Text>
                        </Pressable>

                        <Pressable
                        style={[adminStyle.button, adminStyle.buttonClose]}
                        onPress={()=> props.handleRequestClose()}
                        >
                            <Text style={adminStyle.textStyle}>No, close.</Text>
                        </Pressable>
                    </View>
                </View>
            </Modal>
    );
}

