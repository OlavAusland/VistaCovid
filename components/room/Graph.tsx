import { Dimensions, Platform, TouchableOpacity } from 'react-native';
import { LineChart } from 'react-native-chart-kit';
import { GraphData } from '../../domain/RoomType';

export type GraphProps = {
    data: GraphData[] | undefined,
    name: string,
    color: string
}

export const LineGraph = (props: GraphProps) => {

    const xData = props.data?.map((res: GraphData) => {return res.time.toString()});
    const yData = props.data?.map((res: GraphData) => {return res.value});

    return(
        <TouchableOpacity>
            <LineChart
            data={{
                labels: [],
                datasets: [
                    {
                        data: yData? yData.slice(-15) : [],
                    }
                ],
                legend: [props.name]
            }}
            width={Platform.OS === 'android' ? Dimensions.get('window').width / 1.1 : 375} // from react-native
            height={200}
            xAxisLabel='s'
            yAxisInterval={1} // optional, defaults to 1
            chartConfig={{
                useShadowColorFromDataset:true,
                fillShadowGradientOpacity:0.5,
                backgroundGradientFrom: props.color,
                backgroundGradientTo: props.color,
                decimalPlaces: 1, // optional, defaults to 2dp
                color: () => '#C1E8FD',
                labelColor: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
                style: {
                    borderRadius: 16
            },
            propsForDots: {
                r: "2",
                strokeWidth: "4",
                stroke: "#000000"
            }
            }}
            style={{
                marginVertical: 8,
                borderRadius: 16,
            }}
            fromZero={false}
        />
        </TouchableOpacity>
    )
}
