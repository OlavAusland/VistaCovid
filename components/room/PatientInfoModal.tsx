import { Text, View, TextInput, Pressable, Modal, ScrollView } from 'react-native';
import { useEffect, useState } from 'react';
import { getPatient as folkeregisterGetPatient} from '../../api/folkeregisterModelAPI';
import {getPatient as firebaseGetPatient} from '../../api/firebaseAPI';
import { FolkeregisterPerson, Patient } from '../../domain/PatientType';
import { patientInfoStyle } from '../../styles/PatientInfoStyle';
import { PatientInfoModalProps } from '../../domain/patientInfoType';


export const PatientInfoModal = (props: PatientInfoModalProps) => {
    const [patient, setPatient] = useState<FolkeregisterPerson | Patient>();
    const [error, setError] = useState<Error | undefined>(undefined);

    useEffect(() => {
        folkeregisterGetPatient(props.fnr).then(patient => {
            setPatient(patient);
        }).catch((err) =>{setError(err);});

        if(!patient){
            firebaseGetPatient(props.fnr).then(patient => {
                setPatient(patient);
            }).catch((err) =>{setError(err);});
        }
        
    }, []);

    useEffect(() => {
        console.log(patient);
    }, [patient]);

/* 
    if (error) {
        return <View>
            <Text>{error.message}</Text>
            <Text>Can't find patient info</Text>
        </View>
    } */

    return (

        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            visible={props.modalVisible}
            onRequestClose={() => props.handleRequestClose()}
            testID="patientInfoModal"
        >
            <View style={{ top: '20%', bottom: '15%' }}>
                <View style={patientInfoStyle.container}>
                    <Text style={{ fontSize: 30, fontWeight: '200', marginTop: 10, marginBottom: 10 }}>Patient</Text>
                    <View style={patientInfoStyle.text}>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Lastname: </Text>
                            <Text testID='lastname'>{patient?.lastname}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Firstname: </Text>
                            <Text testID='firstname'>{patient?.firstname} {patient?.midlename}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>SSN: </Text>
                            <Text testID='ssn'>{patient?.ssn}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Gender: </Text>
                            <Text testID='gender'>{patient?.gender}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Address: </Text>
                            <Text  testID='address'>{patient?.address} {patient?.housenumber}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>City: </Text>
                            <Text  testID='city'>{patient?.city}</Text>
                        </Text>
                        <Text style={{ fontSize: 15, marginBottom: 10 }}>
                            <Text style={{ fontWeight: 'bold' }}>Co-Address: </Text>
                            <Text  testID='co.adress'>{patient?.coAddress}</Text>
                        </Text>
                    </View>
                    <Pressable
                        onPress={() => props.handleRequestClose()} >
                        <View style={patientInfoStyle.button}>
                            <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Close</Text>
                        </View>
                    </Pressable>
                </View>
            </View>
        </Modal>
    );
}

