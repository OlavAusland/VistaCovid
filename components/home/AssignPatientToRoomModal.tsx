import { Text, View, TextInput, Pressable, Modal, TouchableOpacity } from 'react-native';
import { useEffect, useState } from 'react';
import { getPatient } from '../../api/folkeregisterAPI';
import { FolkeregisterPerson } from '../../domain/PatientType';
import { assignPatientStyle } from '../../styles/AssignPatientStyle';
import { addPatient, getAvailableRooms } from '../../api/firebaseAPI';
import { Room } from '../../domain/RoomType';
import { addPatientToRoom } from '../../api/firebaseAPI';
import { DropDownType, ItemType } from '../../domain/DropDownType';
import {ExistingPatient} from './patientExist'
import {NewPatient} from './newPatient'
import { AssignPatientModalProps } from '../../domain/AssignPatietTypes';


export const AssignPatientModal = (props: AssignPatientModalProps) => {

    const [patient, setPatient] = useState<FolkeregisterPerson>();
    const [search, setSearch] = useState<string>("");
    const [dropdown, setDropdown] = useState<DropDownType>({ open: false, value: "0", items: [], label: "" });
    const [error, setError] = useState<string>('');
    const [newPatient, setNew] = useState<boolean>(false);

    useEffect(() => {
        getAvailableRooms().then((room: Room[]) => {
            setDropdown(prev => ({ ...prev, items: [] }));
            room.forEach((room: Room) => {
                const item: ItemType = { label: room.id, value: room.id, };
                setDropdown(prev => ({ ...prev, items: [...prev.items, item] }));
            })
        })
    }, [props.modalVisible]);

    const handleSearch = () => {
        if (search.length > 0) {
            const fnrRegex = new RegExp(/^(0[1-9]|[1-2][0-9]|31(?!(?:0[2469]|11))|30(?!02))(0[1-9]|1[0-2])\d{7}$/g);
            const isFnr = fnrRegex.test(search);

            if (!isFnr) {
                setError('Invald SSN') 
            }
            getPatient(search).then(result => {
                setPatient(result)
            }).catch(err => { setError(err.message) });
        }
    }

    const handleAddPatient = () => {
        if (patient && dropdown.label) {
            addPatientToRoom(dropdown.label, patient.ssn);
            setEmpty();
        }
    }
    const handleNewPatient = () => {
        if (patient && dropdown.label) {
            console.log(patient)
            addPatient(patient)
            addPatientToRoom(dropdown.label, patient.ssn);
            setEmpty();
        }
    }

    const setEmpty = () => {
        props.handleRequestClose();
        setPatient(undefined);
        setSearch("");
        setNew(false)
        setDropdown({ open: false, value: "0", items: [], label: "" });
        setError('')
    }

    const handleNew = () => {
       setNew(true);
    }

    return (

        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            visible={props.modalVisible}
            onRequestClose={() => props.handleRequestClose()}
        >
            <View style={{ top: '5%'}}>
                <View style={assignPatientStyle.container}>
                    <Text style={{ fontSize: 40, marginBottom: 60 }}>Admit patient</Text>
                    {newPatient? 
                    <NewPatient 
                        setPatient={setPatient} 
                        patient={patient}
                        setSearch={setSearch} 
                        handleSearch={handleSearch} 
                        dropdown={dropdown} 
                        setDropdown={setDropdown} 
                        handleRequestClose ={props.handleRequestClose}
                        handleNewPatient= {handleNewPatient}
                        setNew ={setNew}
                        setError={setError}
                       /> 
                        
                    :<ExistingPatient 
                        patient={patient} 
                        setSearch={setSearch} 
                        handleSearch={handleSearch} 
                        dropdown={dropdown} 
                        setDropdown={setDropdown}  
                        handleAddPatient={handleAddPatient}
                        handleNew ={handleNew}
                        error={error}
                        setEmpty={setEmpty}/>
                        }
                </View>
            </View>
        </Modal>
    );
}

