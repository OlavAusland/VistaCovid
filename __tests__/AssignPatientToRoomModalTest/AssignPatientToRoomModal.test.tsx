import {AssignPatientModal} from '../../components/home/AssignPatientToRoomModal';
import { PatientInfoModal } from '../../components/room/PatientInfoModal';
import { render, waitFor } from '@testing-library/react-native';
import React from 'react';

test('renders correctly', async() => {
  const tree = render(<AssignPatientModal modalVisible={true} handleRequestClose={jest.fn()} user={}/>);
  await waitFor(() => {
    expect(tree.toJSON()).toMatchSnapshot();
  });
}); 