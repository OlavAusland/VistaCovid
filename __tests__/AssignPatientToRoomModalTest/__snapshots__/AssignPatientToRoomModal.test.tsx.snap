// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  statusBarTranslucent={true}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "top": "5%",
      }
    }
  >
    <View
      style={
        Object {
          "alignContent": "center",
          "alignSelf": "center",
          "backgroundColor": "#9dd4fb",
          "elevation": 5,
          "height": "100%",
          "padding": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "width": "50%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 40,
            "marginBottom": 60,
          }
        }
      >
        Admit patient
      </Text>
      <View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 20,
                "marginTop": 40,
                "paddingBottom": 5,
              }
            }
          >
            Patient:
          </Text>
          <View
            style={
              Object {
                "borderRadius": 5,
                "flexDirection": "row",
                "marginBottom": 10,
                "width": "100%",
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              onChangeText={[Function]}
              placeholder="SSN"
              rejectResponderTermination={true}
              style={
                Object {
                  "backgroundColor": "white",
                  "flex": 3,
                  "height": 40,
                  "paddingLeft": 10,
                  "width": "100%",
                }
              }
              underlineColorAndroid="transparent"
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "flex": 1,
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  Object {
                    "backgroundColor": "#0274a1",
                    "borderRadius": 10,
                    "height": 40,
                    "marginLeft": 10,
                    "width": 70,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "alignSelf": "center",
                      "color": "white",
                      "fontSize": 15,
                      "marginTop": 10,
                    }
                  }
                >
                  Search
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Object {
                "fontSize": 20,
                "marginTop": 25,
                "paddingBottom": 5,
              }
            }
          >
            Room:
          </Text>
          <View
            style={
              Object {
                "backgroundColor": "white",
              }
            }
          >
            <View
              onLayout={[Function]}
              style={
                Array [
                  Object {
                    "justifyContent": "center",
                  },
                  Object {
                    "height": 30,
                    "marginBottom": 10,
                    "marginTop": 10,
                    "paddingLeft": 10,
                    "paddingRight": 10,
                  },
                ]
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "height": 35,
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text />
                <Text
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "fontSize": 16,
                      },
                      Object {
                        "fontSize": 16,
                      },
                      Object {},
                    ]
                  }
                >
                  Select room
                </Text>
                <Image
                  source={1}
                  style={
                    Array [
                      Object {
                        "height": 20,
                        "width": 20,
                      },
                      Object {
                        "tintColor": "gray",
                      },
                      Object {
                        "height": 20,
                        "width": 20,
                      },
                    ]
                  }
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          <Text>
            Responsible: 
          </Text>
        </View>
        <View
          style={
            Object {
              "alignContent": "space-between",
              "flexDirection": "row",
              "marginTop": 180,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "#0274a1",
                  "borderRadius": 10,
                  "height": 40,
                  "marginRight": 20,
                  "marginTop": 30,
                  "width": 150,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignSelf": "center",
                    "color": "white",
                    "fontSize": 20,
                  }
                }
              >
                Add
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "#0274a1",
                  "borderRadius": 10,
                  "height": 40,
                  "marginRight": 20,
                  "marginTop": 30,
                  "width": 150,
                }
              }
            >
              <Text
                style={
                  Object {
                    "alignSelf": "center",
                    "color": "white",
                    "fontSize": 20,
                  }
                }
              >
                Cancel
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
